### ISSUE TITLE
Add more information to the nag about auto_update environment variables

### ISSUE THREAD

**Issue by ulucs** [Homebrew/brew#16481]
### Verification

- [X] This issue's title and/or description do not reference a single formula e.g. `brew install wget`. If they do, open an issue at https://github.com/Homebrew/homebrew-core/issues/new/choose instead.

### Provide a detailed description of the proposed feature

Remove the following nag:

``` 
==> You have set:
  HOMEBREW_NO_AUTO_UPDATE
but we have dramatically sped up and fixed many bugs in the way we do Homebrew updates since.
Please consider unsetting these and tweaking the values based on the new behaviour.
```

### What is the motivation for the feature?

1. I have set the environment variable for a reason
2. The mentioned "changes" are very vague, there is no way of me finding out how the changes effect my use case
3. In the blame, I found out that the changes were made before I set the environment variable, so the changes are moot for me either way. No date is specified in the nag text.
4. I do not like being nagged about making changes that will make homebrew more frustrating to use for me.

### How will the feature be relevant to at least 90% of Homebrew users?

Better user experience since they are not nagged. They will also not mistakenly revert their environment variable which would make them go through the update process they have specifically tried to avoid.

### What alternatives to the feature have been considered?

- Adding a link to how homebrew updates have changed, so users can make a more informed decision
- Not showing the nag text if the user has set the environment variable after the changes have been made

**Comment by MikeMcQuaid** [comment]
> 1. I have set the environment variable for a reason

What reason?

> 4\. I do not like being nagged about making changes that will make homebrew more frustrating to use for me.

On the flip side, we do not like getting issues file from Homebrew users due to them setting custom configuration without which they would not have issues.

> * Adding a link to how homebrew updates have changed, so users can make a more informed decision

We do display links to release notes when you run `brew update`.

**Comment by ulucs** [comment]
> What reason?

brew decided to upgrade all my packages when I was trying to install one package (ffmpeg), it took one hour (thank god for my fast internet, can't imagine how it would go if it was slower) and broke dependencies in my elixir repos thanks to updating OTP without my consent

> On the flip side, we do not like getting issues file from Homebrew users due to them setting custom configuration without which they would not have issues.

Sure, `brew doctor` is a very acceptable place for this nag. You could move it there. Being outside of the supported path is fine with me.

> We do display links to release notes when you run brew update.

Anything I have missed here? 

```
➜  deploy git:(main) brew update && brew upgrade pulumi
==> You have set:
  HOMEBREW_NO_AUTO_UPDATE
but we have dramatically sped up and fixed many bugs in the way we do Homebrew updates since.
Please consider unsetting these and tweaking the values based on the new behaviour.


Updated 7 taps (homebrew/cask-versions, homebrew/cask-fonts, homebrew/services, mongodb/brew, borkdude/brew, homebrew/core and homebrew/cask).
==> New Formulae
c3c                           helm-ls                       rattler-build                 steamguard-cli                zigmod                        zipkin
==> New Casks
font-noto-znamenny-musical-notation          font-protest-revolution                      font-protest-strike                          taccy
font-protest-guerrilla                       font-protest-riot                            shadow-bot                                   znote
==> Outdated Formulae
aom                 colima              gettext             jasper              libnghttp2          mbedtls             openssl@1.1         pympress            sqlite
assimp              coreutils           ghostscript         jmeter              libomp              meson               openssl@3           python@3.10         svt-av1
at-spi2-core        curl                glib                jpeg                libraw              moreutils           p11-kit             python@3.11         tcl-tk
azure-cli           dav1d               gmp                 jpeg-turbo          librist             mpfr                pandoc              python@3.9          tmux
babashka            docker              gnu-getopt          jpeg-xl             libsodium           mpg123              pandocomatic        qemu                tree
bdw-gc              dotnet              gnutls              jq                  libsoup             nss                 pdftk-java          qt                  unbound
berkeley-db         elixir              go                  leptonica           libssh              nvm                 perl                r                   utf8proc
ca-certificates     erlang              gpgme               libavif             libvmaf             oniguruma           poppler             rav1e               watch
cargo-c             faad2               grep                libde265            libx11              openapi-generator   postgresql@15       readline            wxwidgets
cjson               fdk-aac             gstreamer           libffi              lima                openblas            protobuf            redis               xz
clojure             ffmpeg              gtk+3               libgcrypt           little-cms2         openjdk             pulumi              ruby                yabai
cmake               fontconfig          gtk4                libheif             llvm                openjdk@11          py3cairo            rust                z3
cmocka              frei0r              harfbuzz            libidn              luajit              openjdk@17          pygments            sdl2                zeromq
cocoapods           gd                  imagemagick         libksba             luajit-openresty    opensearch          pygobject3          shared-mime-info
==> Outdated Casks
anydesk                       aws-vault                     kodi                          pgadmin4                      rstudio                       yacreader
audacity                      eloston-chromium              mos                           qbittorrent                   temurin17

You have 125 outdated formulae and 11 outdated casks installed.
You can upgrade them with brew upgrade
or list them with brew outdated.
==> Upgrading 1 outdated package:
pulumi 3.73.0 -> 3.101.1
==> Downloading https://ghcr.io/v2/homebrew/core/pulumi/manifests/3.101.1
############################################################################################################################################################################## 100.0%
==> Fetching pulumi
==> Downloading https://ghcr.io/v2/homebrew/core/pulumi/blobs/sha256:4e71ce0bbea30cd965536ccfe63abd732e6eff3dd965a13039026793f3e4eb38
############################################################################################################################################################################## 100.0%
==> Upgrading pulumi
  3.73.0 -> 3.101.1 

==> Pouring pulumi--3.101.1.arm64_sonoma.bottle.tar.gz
==> Caveats
zsh completions have been installed to:
  /opt/homebrew/share/zsh/site-functions
==> Summary
🍺  /opt/homebrew/Cellar/pulumi/3.101.1: 21 files, 331.6MB
==> Running `brew cleanup pulumi`...
Removing: /opt/homebrew/Cellar/pulumi/3.73.0... (21 files, 292.6MB)
```

**Comment by MikeMcQuaid** [comment]
> brew decided to upgrade all my packages when I was trying to install one package (ffmpeg), it took one hour (thank god for my fast internet, can't imagine how it would go if it was slower) and broke dependencies in my elixir repos thanks to updating OTP without my consent

Setting `HOMEBREW_NO_INSTALLED_DEPENDENTS_CHECK` may be a better fit for your use-case here.

> > We do display links to release notes when you run brew update.
> 
> Anything I have missed here?

There was not a new release during that particular `brew update` run. https://brew.sh/blog/ provides the relevant major/minor release notes.

**Comment by ulucs** [comment]
> Setting `HOMEBREW_NO_INSTALLED_DEPENDENTS_CHECK` may be a better fit for your use-case here.

I could be entirely wrong here, but this is my current model of how homebrew handles dependencies:
1. There is a tree of dependencies for each formula, adding up to a graph (during my update, ffmpeg went down to one of jpeg-turbo or libpng and came back up to elixir)
2. The graph is evergreen, meaning dependency versions aren't pinned. The downloaded ffmpeg was only confirmed to be compatible with the newest jpeg-turbo or libpng which was confirmed to be compatible with the newest elixir. Therefore any library that is touched by the graph needs to be updated for compatibility.
3. The graph is sometimes broken, so running `brew update` will fix issues by updating to the latest graph and upgrading dependencies accordingly.
4. HOMEBREW_NO_AUTO_UPDATE skips updating to the newest graph, so I'm out of luck if my current graph is broken.
5. HOMEBREW_NO_INSTALLED_DEPENDENTS_CHECK skips checking dependencies, so I might have incompatibility errors if the dependencies are outdated.

So, if my graph is not broken, I can skip updating with HOMEBREW_NO_AUTO_UPDATE and install anything without any worries. Everything in the graph will be compatible with each other: the newly installed ffmpeg will work with the libpng which already works with elixir. But if I set HOMEBREW_NO_INSTALLED_DEPENDENTS_CHECK, any library I install will be the newest one, and will not necessarily work with the older installed dependencies. If ffmpeg does not play well with one of  the older dependencies for some reason, I'll have to upgrade all of them, which will have ripple effects through all the libraries I have installed with brew.

I understand that brew does not want to get involved with versioned dependencies, so I just want my decision to stay on an earlier dependency graph to be respected. But as I found out that it was a one-time message, it's now just not really important. Maybe add that information to the nag?

~~I also found the update post at https://brew.sh/2023/07/20/homebrew-4.1.0/ , but I still don't know how the update behavior is changed compared to before.~~ Finally found the relevant blog post: https://brew.sh/2023/02/16/homebrew-4.0.0/ . I still believe the nag would be more useful if it contained the following info: "The most significant change since 3.6.0 enables significantly faster Homebrew-maintained tap updates by migrating from Git-cloned taps to JSON downloads." Then I could confidently drop the env variable if my concern was the slow update rather than moving between dependency graphs.

**Comment by ulucs** [comment]
So in short, I propose the nag message would be more informative/useful and less annoying if it has the following additional information:

- That it is a one-time message
- The main change is the speed-up obtained from moving to the JSON API compared to git-clones.

I can file a PR if you think this is agreeable.

**Comment by MikeMcQuaid** [comment]
> So, if my graph is not broken, I can skip updating with HOMEBREW_NO_AUTO_UPDATE and install anything without any worries. Everything in the graph will be compatible with each other: the newly installed ffmpeg will work with the libpng which already works with elixir.

This mostly works as long as you're not including third-party taps or taps that have been tapped outside of the last auto-update window.

> * HOMEBREW_NO_AUTO_UPDATE skips updating to the newest graph, so I'm out of luck if my current graph is broken.
> * HOMEBREW_NO_INSTALLED_DEPENDENTS_CHECK skips checking dependencies, so I might have incompatibility errors if the dependencies are outdated.

correct 👍🏻 

> * That it is a one-time message

We don't note this on any other one-time messages so I don't think we should here.

> * The main change is the speed-up obtained from moving to the JSON API compared to git-clones, rather than the upgrade surface being smaller

I would agree with this 👍🏻 

> I can file a PR if you think this is agreeable.

Thanks! Let me get back to you; we may just remove this message, I want to ask some other maintainers their thoughts.

@Homebrew/brew and particularly @Bo98: do you think we still need this `installfromapimessage` one-time message or shall we drop it? Alternatively, do we still need checks for `HOMEBREW_NO_AUTO_UPDATE` or `HOMEBREW_API_AUTO_UPDATE_SECS` here or shall we update the messaging?

**Comment by apainintheneck** [comment]
IIRC this message was added in brew v4.1 which got released on July 20, 2023 and brew v4.2 got released last month on December 18, 2023. That means this message has been active for around 6 months. Anyone that this message was meant to reach has probably seen it by now. I'm fine with dropping it.

**Comment by MikeMcQuaid** [comment]
@apainintheneck Thanks for the feedback! I mostly agree; I still think it's worth nudging people to not have `HOMEBREW_NO_INSTALL_FROM_API` set but the other variable checks have been dropped as well as noting (as requested by @ulucs) it's a one-time message: https://github.com/Homebrew/brew/pull/16506
